### PR TITLE
Allocate memory space for slice in advance

### PR DESCRIPTION
**Pull request by Fish-pro**
**Please provide a description of this PR:**

I recently did some local testing and found that allocating memory to slice in advance greatly reduced the execution time of the program. So this PR is to help optimize memory allocation and improve the execution efficiency of the program

```golang
func AppendVar() []int {
	var res []int
	for i := 0; i < 10; i++ {
		res = append(res, i)
	}
	return res
}

func AppendPreAllocate() []int {
	res := make([]int, 0, 10)
	for i := 0; i < 10; i++ {
		res = append(res, i)
	}
	return res
}

func AppendPreAllocateIndex() []int {
	res := make([]int, 10)
	for i := 0; i < 10; i++ {
		res[i] = i
	}
	return res
}

func BenchmarkAppendVar(b *testing.B) {
	for i := 0; i < b.N; i++ {
		AppendVar()
	}
}

func BenchmarkAppendPreAllocate(b *testing.B) {
	for i := 0; i < b.N; i++ {
		AppendPreAllocate()
	}
}

func BenchmarkAppendPreAllocateIndex(b *testing.B) {
	for i := 0; i < b.N; i++ {
		AppendPreAllocateIndex()
	}
}
```

image

### CODE REVIEW

File: istioctl/pkg/util/formatting/formatter.go [hunk 1]
```diff
@@ -62,9 +62,9 @@ func Print(ms diag.Messages, format string, colorize bool) (string, error) {
 }
 
 func printLog(ms diag.Messages, colorize bool) string {
-	var logOutput []string
-	for _, m := range ms {
-		logOutput = append(logOutput, render(m, colorize))
+	logOutput := make([]string, len(ms))
+	for i, m := range ms {
+		logOutput[i] = render(m, colorize)
```
Comment (zirain) [inline istioctl/pkg/util/formatting/formatter.go:67]:
```suggestion
	logOutput := make([]string, 0, len(ms))
	for i, m := range ms {
		logOutput = append(logOutput, render(m, colorize))
```

Comment (Fish-pro) [reply istioctl/pkg/util/formatting/formatter.go:67]: @zirain According to my tests, writing values using an index is more efficient than writing values using append when slice determines the length. Are there any other special cases?
image

image

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: istioctl/pkg/util/formatting/formatter.go (hunk 1 above)
Comment (Fish-pro) [reply istioctl/pkg/util/formatting/formatter.go:67]: > I am fine with pre-allocating, but please use `append`. Microbenchmarks on this nanosecond scale are not applicable to real world

@howardjohn Thanks~, I got it

**Comment by Fish-pro** [comment]
@zirain @howardjohn Thanks for guidance, it has been revised

**Review by zirain — APPROVED** [review]
Thanks!

**Review by hanxiaop — APPROVED** [review]

**Review by ymesika — APPROVED** [review]
Nice benchmarking work.

File: pilot/pkg/serviceregistry/kube/controller/controller_test.go [hunk 2]
```diff
@@ -2472,7 +2472,7 @@ func TestWorkloadInstanceHandlerMultipleEndpoints(t *testing.T) {
 	// we should have the pod IP and the workload Entry's IP in the endpoints..
 	// the first endpoint should be that of the k8s pod and the second one should be the workload entry
 
-	var gotEndpointIPs []string
+	gotEndpointIPs := make([]string, 0, len(ev.Endpoints))
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller_test.go:2475]: nit: I don't mind this but in the future, IMO, readability/simplicity > performance in tests

**Review by howardjohn — APPROVED** [review]

**Pull request merged**